src/wallet/wallet.cpp

```diff
@@ -2880,21 +2880,31 @@ CBitcoinAddress CWallet::ParseIntoAddress(const CTxDestination& dest, const std:
     return CBitcoinAddress(dest, addrType);
 }
 
+/**
+ * TODO: If we ever add new functionality that modifies the purpose of an address book entry
+ * outside of the below situations, we need to also ensure those situations are considered here
+ */
+bool CWallet::UpdatablePurpose(const std::string& strPurpose1, const std::string& strPurpose2)
+{
+    return (strPurpose1 == AddressBook::AddressBookPurpose::DELEGABLE && strPurpose2 == AddressBook::AddressBookPurpose::DELEGATOR) ||
+           (strPurpose1 == AddressBook::AddressBookPurpose::DELEGATOR && strPurpose2 == AddressBook::AddressBookPurpose::DELEGABLE);
+}
+
 bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& strName, const std::string& strPurpose)
 {
     bool fUpdated = HasAddressBook(address);
+    const bool fUpdatePurpose = !fUpdated || UpdatablePurpose(mapAddressBook.at(address).purpose, strPurpose);
     {
         LOCK(cs_wallet); // mapAddressBook
         mapAddressBook[address].name = strName;
-        if (!strPurpose.empty()) /* update purpose only if requested */
-            mapAddressBook[address].purpose = strPurpose;
+        if (!strPurpose.empty() || fUpdatePurpose) mapAddressBook[address].purpose = strPurpose;
     }
     NotifyAddressBookChanged(this, address, strName, ::IsMine(*this, address) != ISMINE_NO,
         strPurpose, (fUpdated ? CT_UPDATED : CT_NEW));
     if (!fFileBacked)
         return false;
-    std::string addressStr = ParseIntoAddress(address, strPurpose).ToString();
-    if (!strPurpose.empty() && !CWalletDB(strWalletFile).WritePurpose(addressStr, strPurpose))
+    std::string addressStr = ParseIntoAddress(address, (strPurpose.empty() && fUpdated) ? mapAddressBook.at(address).purpose : strPurpose).ToString();
+    if ((!fUpdated || fUpdatePurpose) && !CWalletDB(strWalletFile).WritePurpose(addressStr, strPurpose))
```

Comment:
To be totally sure over the write purpose process, would use the `IsMine` method call that is few lines above to verify that the address is really from this wallet (no need to do any extra processing, just use what we already have here). And throw an exception if the wallet, for whatever reason, is trying to store an external address with the receive/cold_staking purpose.